aten/src/ATen/native/cuda/ForeachTernaryOp.cu

```diff
@@ -0,0 +1,249 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Dispatch.h>
+#include <ATen/native/ForeachUtils.h>
+#include <ATen/native/cuda/ForeachFunctors.cuh>
+#include <ATen/native/cuda/MultiTensorApply.cuh>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_foreach_lerp_native.h>
+
+#include <ATen/ops/empty_like_native.h>
+#endif
+
+namespace at { namespace native {
+
+template <typename scalar_t>
+inline C10_HOST_DEVICE scalar_t lerp_(const scalar_t start, const scalar_t end, const scalar_t weight) {
+  using opmath_t = at::opmath_type<scalar_t>;
+  opmath_t start_val = start;
+  opmath_t end_val = end;
+  opmath_t weight_val = weight;
+  // Conditional for better numeric. This has been discussed in
+  // https://github.com/pytorch/pytorch/pull/18871
+  return (std::abs(weight_val) < 0.5)
+      ? start_val + weight * (end_val - start_val)
+      : end_val - (end_val - start_val) * (opmath_t{1} - weight_val);
+}
+
+// Ref: [GPU Pro Tip: Lerp Faster in C++](https://developer.nvidia.com/blog/lerp-faster-cuda/)
+template<>
+inline C10_HOST_DEVICE float lerp_(const float start, const float end, const float weight) {
+  return fma(weight, end, fma(-weight, start, start));
```

Comment:
this is different than generic implementation above, and than eager implementation, do we want this divergence? Each of the branches in the generic implementation can be rewritten with 2 fmas like here, so maybe we should just do that? For fixed weight there would be no branch divergence

Reply:
I wanted to make sure the precision of this fma based implementation first by comparing foreach_lerp against at::native::lerp before introducing integrating to the at::native::lerp.
I think I've overlooked some tests which confirms that parity of at::natie::lerp_cpu and at::native::lerp_cuda

Reply:
so, my question is, would you mind if I update ATen/native/cuda/Lerp.cu to use this fma based logic for certain dtyps?

Reply:
I don't mind fma logic, however, we should keep weight > 0.5 comparison and have 2 fma branches.